src/main/java/duke/TaskStorage.java

```diff
@@ -0,0 +1,133 @@
+import java.util.Scanner;
+
+import duke.task.Deadline;
+import duke.task.Event;
+import duke.task.Task;
+import duke.task.ToDos;
+
+public class TaskStorage {
+    static final String DATA_FILE_PATH = "./data/tasks.txt";
+    static final String SPACE_SEPARATOR = "\\|";
+
+    private File dataFile;
+    private ArrayList<Task> tasks = new ArrayList<>();
+
+    public TaskStorage() throws IOException {
+        File dataFile = new File(DATA_FILE_PATH);
+        File dataDir = new File(dataFile.getParent());
+
+        if (!dataDir.exists()) {
+            dataDir.mkdirs();
+        }
+
+        if (!dataFile.exists()) {
+            dataFile.createNewFile();
+        }
+
+        this.dataFile = dataFile;
+    }
+
+    public ArrayList<Task> tasksFromFile() {
+        ArrayList<Task> extractedTasks = new ArrayList<>();
+
+        try {
+            Scanner scanner = new Scanner(this.dataFile);
+
+            while(scanner.hasNextLine()) {
+                Task task = readTask(scanner.nextLine());
+                extractedTasks.add(task);
+            }
+        } catch (FileNotFoundException e) {
+            System.out.println("\tI cannot open the file!");
+        } catch (DukeException e) {
+            System.out.println("\tWrong task format");
+        }
+
+        tasks = extractedTasks;
+        return extractedTasks;
+    }
+
+    public void storageAddTask(Task task) throws IOException {
+        addTask(task);
+    }
+
+    public Task storageDeleteTask(int taskNumber) throws IOException, IndexOutOfBoundsException {
+        return deleteTask(taskNumber);
+    }
+
+    public void storageSetTaskDone(int taskNumber) throws IOException, IndexOutOfBoundsException {
+        setTaskDone(taskNumber);
+    }
+
+    private void taskToFile(ArrayList<Task> tasks) throws IOException {
+        String data = "";
+        FileWriter fw = new FileWriter(DATA_FILE_PATH);
+
+        for (Task t : tasks) {
+            data += taskToString(t) + System.lineSeparator();
+        }
+
+        fw.write(data);
+        fw.close();
+    }
+
+    private String taskToString(Task t) {
+        if (t instanceof ToDos) {
+            ToDos todoTask = (ToDos) t;
+            return String.format("todo|%s|%b", todoTask.getTaskDescription(),
+                    todoTask.isDone());
+        } else if (t instanceof Event) {
+            Event eventTask = (Event) t;
+            return String.format("event|%s|%s|%b", eventTask.getTaskDescription(),
+                    eventTask.getDate(), eventTask.isDone());
+        } else if (t instanceof  Deadline) {
+            Deadline deadlineTask = (Deadline) t;
+            return String.format("deadline|%s|%s|%b", deadlineTask.getTaskDescription(),
+                    deadlineTask.getDate(), deadlineTask.isDone());
+        } else {
+            return null;
+        }
+    }
+
+    private Task readTask(String string) throws DukeException {
+        String[] detail = string.split(SPACE_SEPARATOR);
```

Comment:
`details` would better reflect the plural parts. Also the input parameter could be better named as `taskString` or `taskDetailsString` to make it clearer